tests/apps/test_similarity.py

```diff
@@ -222,217 +223,474 @@ def test_event_cardinality(photons, max_count, modes, expected):
+
+    def test_invalid_loss(self):
+        """Test if function raises a ``ValueError`` when the loss parameter is specified outside
+        of range."""
+        g = nx.complete_graph(5)
+        with pytest.raises(ValueError, match="Loss parameter must take a value between zero and one"):
+            similarity.prob_event_exact(g, 2, 2, loss=2)
+
+    def test_invalid_photon_number(self):
+        """Test if function raises a ``ValueError`` when a photon number below zero is specified"""
+        g = nx.complete_graph(5)
+        with pytest.raises(ValueError, match="Photon number must not be below zero"):
+            similarity.prob_event_exact(g, -1, 2)
+
+    def test_low_count(self):
+        """Test if function raises a ``ValueError`` if ``max_count_per_mode`` is negative."""
+        g = nx.complete_graph(5)
+        with pytest.raises(ValueError, match="Maximum number of photons per mode must be non-negative"):
+            similarity.prob_event_exact(g, 2, max_count_per_mode=-1)
+
+    def test_prob_vacuum_event(self):
+        """Tests if the function gives the right probability for an event with zero photons when
+        the GBS device has been configured to have zero mean photon number."""
+        graph = nx.complete_graph(5)
+        assert similarity.prob_event_exact(graph, 0, 0, 0) == 1.0
+
+    @pytest.mark.parametrize("k", [3, 5, 7, 9])
+    @pytest.mark.parametrize("nmax", [1, 2])
+    def test_odd_photon_numbers(self, k, nmax):
+        """Test if function returns zero probability for odd number of total photons."""
+        graph = nx.complete_graph(10)
+        assert similarity.prob_event_exact(graph, k, nmax) == 0.0
+
+    def test_correct_result_returned(self, monkeypatch):
+        """Tests if the call to _get_state function is performed correctly and probabilities over all
+        constituent orbits of the given event are summed correctly. The test monkeypatches the fock_prob
+        function so that the probability is the same for each sample permutation of all constituent orbits
+        and is equal to 1/8. For a 4-mode graph, an event with ``photon_number = 2``, and
+        ``max_count_per_mode = 1`` contains orbit [1, 1] which has 6 possible sample permutations."""
+        graph = nx.complete_graph(4)
+        with monkeypatch.context() as m:
+            m.setattr("strawberryfields.backends.gaussianbackend.states.GaussianState.fock_prob", lambda *args, **kwargs: 1/8)
+            assert similarity.prob_event_exact(graph, 2, 1) == 6/8
+
+    def test_known_result(self):
+        """Tests if the probability for detecting multiple photons in one mode, orbit [2], is
+        correctly reproduced to be close to zero for a small graph and a small mean photon number.
+        To get orbit [2], we use ``p(E_{2,2}) - p(E_{2,1})``. """
+        graph = nx.complete_graph(4)
+        assert np.allclose((similarity.prob_event_exact(graph, 2, 2, 1) - similarity.prob_event_exact(graph, 2, 1, 1)), 0)
+
```

Comment:
Would be good to run black on `test_similarity.py` before merging. It's not required but just keeps things tidy.

Reply:
running black spreads some lines of code over many lines that JM did not like. so, I am manually doing the formatting here.

Reply:
I recommend splitting up into a few separate lines rather than going over line width:
```python
p1 = similarity.prob_event_exact(graph, 2, 2, 1)
p2 = similarity.prob_event_exact(graph, 2, 1, 1)
assert np.allclose(p1 - p2, 0)
```
Let's keep the standard of running `black` for `test_similarity.py`. In general, we shouldn't worry about second guessing the formatting of `black` (the idea is that it saves you thinking about that). However, in some cases the result will be brackets split over quite a few lines - in that case we can define some intermediate variables as above for readability.